### PR TITLE
BREAKING: `PostMeltRequest`, CheckSpendableResponse `

### PR DESCRIPTION
**Pull request by callebtc**
## BREAKING CHANGES

- Model `PostMeltRequest` field `invoice` renamed to `pr` for better consistency.
- Model `GetCheckSpendableResponse` field "spendable" has now an array.

#### `PostMeltRequest` 

Here simply the field "invoice" is renamed to "pr". `PostMeltRequest` is now a JSON of the form

```json
{
  "proofs" : [Proof, ...],
  "pr": <str> 
}
```

#### `CheckSpendableResponse`

This object was very ugly before as it was a dictionary with variable keys. `CheckSpendableResponse` is now a JSON with a fixed key "spendable" and an array of the form 

```json
{
  "spendable": [true, false, ...]
}
```

#### Mints
- [x] Python
- [ ] Golang

#### Wallets
- [x] Python
- [ ] Golang
- [x] Nutstash 
- [ ] LNbits
- [x] cashu-js

### CODE REVIEW

**Comment by gandlafbtc** [comment]
> Model PostMeltRequest field invoice renamed to pr for better consistency.

Is that an internal thing or is the API affected?

**Comment by gandlafbtc** [comment]
cashu-ts is ready: https://github.com/gandlafbtc/cashu-ts/pull/1

File: cashu/core/base.py [hunk 1]
```diff
@@ -138,7 +138,7 @@ class GetMintResponse(BaseModel):
 
 class PostMeltRequest(BaseModel):
     proofs: List[Proof]
-    invoice: str
+    pr: str
```
Comment (callebtc) [inline cashu/core/base.py:141]: This is the relevant change.

File: cashu/core/base.py [hunk 2]
```diff
@@ -163,11 +163,15 @@ class PostSplitResponse(BaseModel):
...
 
 
-class GetCheckSpendableRequest(BaseModel):
+class CheckSpendableRequest(BaseModel):
     proofs: List[Proof]
 
 
-class GetCheckFeesRequest(BaseModel):
+class CheckSpendableResponse(BaseModel):
+    spendable: List[bool]
```
Comment (callebtc) [inline cashu/core/base.py:171]: This is the relevant change.

File: cashu/mint/ledger.py [hunk 3]
```diff
@@ -412,7 +412,7 @@ async def melt(self, proofs: List[Proof], invoice: str):
 
     async def check_spendable(self, proofs: List[Proof]):
         """Checks if all provided proofs are valid and still spendable (i.e. have not been spent)."""
-        return {i: self._check_spendable(p) for i, p in enumerate(proofs)}
+        return [self._check_spendable(p) for p in proofs]
```
Comment (callebtc) [inline cashu/mint/ledger.py:415]: Returns a list now.

File: cashu/mint/router.py [hunk 4]
```diff
@@ -121,17 +122,20 @@ async def melt(payload: PostMeltRequest) -> Union[CashuError, GetMeltResponse]:
...
     summary="Check whether a proof has already been spent",
 )
-async def check_spendable(payload: GetCheckSpendableRequest) -> Dict[int, bool]:
+async def check_spendable(
+    payload: CheckSpendableRequest,
+) -> CheckSpendableResponse:
     """Check whether a secret has been spent already or not."""
-    return await ledger.check_spendable(payload.proofs)
+    spendableList = await ledger.check_spendable(payload.proofs)
+    return CheckSpendableResponse(spendable=spendableList)
```
Comment (callebtc) [inline cashu/mint/router.py:130]: List is then loaded into the model and the model is returned.

File: cashu/wallet/wallet.py [hunk 5]
```diff
@@ -355,11 +356,12 @@ def _check_spendable_include_fields(proofs):
         resp.raise_for_status()
         return_dict = resp.json()
         self.raise_on_error(return_dict)
-        return return_dict
+        spendable = CheckSpendableResponse.parse_obj(return_dict)
+        return spendable
```
Comment (callebtc) [inline cashu/wallet/wallet.py:360]: List is then loaded into the model and the model is returned.

File: cashu/wallet/wallet.py [hunk 6]
```diff
@@ -374,14 +376,14 @@ async def pay_lightning(self, proofs: List[Proof], invoice: str):
         """
         Accepts proofs and a lightning invoice to pay in exchange.
         """
-        payload = PostMeltRequest(proofs=proofs, invoice=invoice)
+        payload = PostMeltRequest(proofs=proofs, pr=invoice)
```
Comment (callebtc) [inline cashu/wallet/wallet.py:379]: Wallet: This is the API change.

File: cashu/wallet/wallet.py [hunk 7]
```diff
@@ -374,14 +376,14 @@ async def pay_lightning(self, proofs: List[Proof], invoice: str):
...
-        payload = PostMeltRequest(proofs=proofs, invoice=invoice)
+        payload = PostMeltRequest(proofs=proofs, pr=invoice)
 
         def _meltrequest_include_fields(proofs):
             """strips away fields from the model that aren't necessary for the /melt"""
             proofs_include = {"id", "amount", "secret", "C", "script"}
             return {
                 "amount": ...,
-                "invoice": ...,
+                "pr": ...,
```
Comment (callebtc) [inline cashu/wallet/wallet.py:386]: Wallet: This is the API change.

File: cashu/wallet/wallet.py [hunk 8]
```diff
@@ -609,10 +611,10 @@ async def invalidate(self, proofs):
         """Invalidates all spendable tokens supplied in proofs."""
         spendables = await self.check_spendable(proofs)
         invalidated_proofs = []
-        for idx, spendable in spendables.items():
+        for i, spendable in enumerate(spendables.spendable):
             if not spendable:
-                invalidated_proofs.append(proofs[int(idx)])
-                await invalidate_proof(proofs[int(idx)], db=self.db)
+                invalidated_proofs.append(proofs[i])
+                await invalidate_proof(proofs[i], db=self.db)
```
Comment (callebtc) [inline cashu/wallet/wallet.py:617]: Wallet: This is the Model change.

**Pull request merged**